packages/contracts-bedrock/src/periphery/monitoring/DisputeMonitorHelper.sol

```diff
@@ -0,0 +1,157 @@
+        // likely to be a common standard case and it'd be nicer if it didn't error.
+        if (_factory.gameCount() == 0) {
+            return new IDisputeGame[](0);
+        }
+
+        // Try to find a suitable start and end index. If startIdx is type(uint256).max then we did
+        // not find any newer games and the creation range start must be after the timestamp of the
+        // latest game. Similarly, if endIdx is type(uint256).max then we did not find any older
+        // games and the creation range end must be before the timestamp of the earliest game. In
+        // either case, we can return an empty array.
+        uint256 startIdx = search(_factory, _creationRangeStart, SearchDirection.NEWER_THAN_OR_EQ);
+        uint256 endIdx = search(_factory, _creationRangeEnd, SearchDirection.OLDER_THAN_OR_EQ);
+        if (startIdx == type(uint256).max || endIdx == type(uint256).max) {
+            return new IDisputeGame[](0);
+        }
+
+        // Additionally, if the end index is less than the start index, then the range is between
+        // two dispute games. We return an empty array in this case.
+        if (endIdx < startIdx) {
+            return new IDisputeGame[](0);
+        }
+
+        // Allocate the array and fill it
+        unresolvedGames_ = new IDisputeGame[](endIdx - startIdx + 1);
+        uint256 unresolvedGameCount = 0;
+        for (uint256 i = startIdx; i <= endIdx; i++) {
+            (,, IDisputeGame game) = _factory.gameAtIndex(i);
+            if (game.resolvedAt().raw() == 0) {
+                unresolvedGames_[unresolvedGameCount] = game;
+                unresolvedGameCount++;
+            }
+        }
+
+        // Clobber the size of the array to return the right size.
+        assembly {
+            mstore(unresolvedGames_, unresolvedGameCount)
+        }
+    }
+
+    /// @notice Searches for a game by timestamp, returning the index of the game that best matches the
+    ///         given search direction.
+    /// @param _factory         The factory of the dispute games.
+    /// @param _targetTimestamp The timestamp to search for.
+    /// @param _direction       The direction to search in (older or newer).
+    /// @return index_          The index of the matching game, if it exists.
+    function search(
+        IDisputeGameFactory _factory,
+        uint256 _targetTimestamp,
+        SearchDirection _direction
+    )
+        public
+        view
+        returns (uint256 index_)
+    {
+        uint256 gameCount = _factory.gameCount();
+        // If there are no games, return max to indicate "not found."
+        if (gameCount == 0) {
+            return type(uint256).max;
+        }
+
+        uint256 left = 0;
+        uint256 right = gameCount - 1;
+
+        // We'll store the candidate here. If it remains max, no suitable game was found.
+        index_ = type(uint256).max;
+
+        while (left <= right) {
```

Comment:
Generally with any search algo like this, I'd prefer a pure algo w/ fuzz tests, and then a specific wrapper ontop, although this does look ok, but it does make it harder to verify

Reply:
Very valid, could make it more generic

Reply:
Since this isn't too critical, I think it's probably fine as a specific function. If we end up needing another similar search function somewhere else I'd be happy to make this generic in the future.